controllers/librarypanel_controller.go

```diff
@@ -0,0 +1,519 @@
+
+	instances, err := GetScopedMatchingInstances(ctx, r.Client, libraryPanel)
+	if err != nil {
+		return fmt.Errorf("fetching instances: %w", err)
+	}
+	for _, i := range instances {
+		instance := i
+		if err := r.removeFromInstance(ctx, &instance, libraryPanel); err != nil {
+			return fmt.Errorf("removing notification template from instance: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (r *GrafanaLibraryPanelReconciler) removeFromInstance(ctx context.Context, instance *v1beta1.Grafana, libraryPanel *v1beta1.GrafanaLibraryPanel) error {
+	log := logf.FromContext(ctx)
+
+	grafanaClient, err := client2.NewGeneratedGrafanaClient(ctx, r.Client, instance)
+	if err != nil {
+		return err
+	}
+
+	updateInstanceRefs := func() error {
+		instance.Status.LibraryPanels = instance.Status.LibraryPanels.Remove(libraryPanel.Namespace, libraryPanel.Name)
+		return r.Client.Status().Update(ctx, instance)
+	}
+
+	uid := libraryPanel.Status.UID
+
+	resp, err := grafanaClient.LibraryElements.GetLibraryElementByUID(uid)
+	if err != nil {
+		var notFound *library_elements.GetLibraryElementByUIDNotFound
+		if errors.As(err, &notFound) {
+			// ensure we update the list of managed panels, otherwise we will have dangling references
+			return updateInstanceRefs()
+		}
+		return fmt.Errorf("fetching library panel from instance %s: %w", instance.Status.AdminUrl, err)
+	}
+
+	var wrappedRes *models.LibraryElementResponse
+	if resp != nil {
+		wrappedRes = resp.GetPayload()
+	}
+	var elem *models.LibraryElementDTO
+	if wrappedRes != nil {
+		elem = wrappedRes.Result
+	}
+
+	switch hasConnections, err := libraryElementHasConnections(grafanaClient, uid); {
+	case err != nil:
+		return fmt.Errorf("fetching library panel from instance %s: %w", instance.Status.AdminUrl, err)
+	case hasConnections:
+		return fmt.Errorf("library panel %s on instance %s has existing connections", uid, instance.Status.AdminUrl) //nolint
+	}
+
+	_, err = grafanaClient.LibraryElements.DeleteLibraryElementByUID(uid) //nolint:errcheck
+	if err != nil {
+		var notFound *library_elements.DeleteLibraryElementByUIDNotFound
+		if !errors.As(err, &notFound) {
+			return err
+		}
+	}
+
+	if elem != nil && elem.Meta != nil && elem.Meta.FolderUID != "" {
+		resp, err := r.DeleteFolderIfEmpty(grafanaClient, elem.Meta.FolderUID)
+		if err != nil {
+			return err
+		}
+		if resp.StatusCode == 200 {
+			log.Info("unused folder successfully removed")
+		}
+		if resp.StatusCode == 432 {
+			log.Info("folder still in use by other resources")
+		}
+	}
+
+	return updateInstanceRefs()
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *GrafanaLibraryPanelReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1beta1.GrafanaLibraryPanel{}).
+		WithEventFilter(ignoreStatusUpdates()).
+		Complete(r)
+}
+
+func (r *GrafanaLibraryPanelReconciler) GetOrCreateFolder(client *genapi.GrafanaHTTPAPI, cr *v1beta1.GrafanaLibraryPanel) (string, error) {
```

Comment:
these folder functions are obvious contenders for extraction b/c i just copied them from the dashboard reconciler.

Reply:
As we don't want to support implicit folder creation, this is no longer needed anyways